src/modules/fw_pos_control/FixedwingPositionControl.cpp

```diff
@@ -986,8 +983,13 @@ FixedwingPositionControl::handle_setpoint_type(const position_setpoint_s &pos_sp
 
 	const float acc_rad = _npfg.switchDistance(500.0f);
 
-	if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION
-	    || pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
+	const bool approaching_vtol_backtransition = _vehicle_status.is_vtol
+			&& pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION && _position_setpoint_current_valid
+			&& pos_sp_next.type == position_setpoint_s::SETPOINT_TYPE_LAND && _position_setpoint_next_valid;
+
+
+	// check if we should switch to loiter but only if we are not expecting a backtransition to happen
+	if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION && !approaching_vtol_backtransition) {
```

Comment:
@KonradRudin The reason I added that is because we have two separate modules (navigator & position controller) which could use different horizontal acceptance radius and therefore not be "in sync" anymore. I agree that in theory now this should not be needed as I assume navigator and the position controller use the same value.
It's a protection in the sense that if we know we are expecting a backtransition, I don't ever want it to loiter.
The only bad case I see is if for some reason the logic in navigator for switching to the next waypoint would not trigger, but that would mean that the vehicle is blown away so much that the distance to the land waypoint does not reach the minimum required.
That could in theory happen when you have extremely strong wind and thus a very low groundspeed. That would make the transition distance very small and in theory it would be possible to "miss" the transition point.
We could add a minimum transition distance to make sure that risk is lower.
What do you think?

Reply:
Well, we need to think of what is the best thing to do when we reached the end waypoint and the back transition has not triggered yet. Before it was loiter, now it the "flower" pattern, both are not good. But i get it, that in theory it should only happen when we are blown away so flying to the pattern might be better. Still hate this obfuscating navigator <-> controller interaction and should really kill current navigator once to make this better. But let's get it in this way.